src/main/java/io/confluent/copycat/hdfs/WAL.java

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright 2015 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ **/
+
+package io.confluent.copycat.hdfs;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathFilter;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.ipc.RemoteException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+
+import io.confluent.copycat.connector.TopicPartition;
+import io.confluent.copycat.errors.CopycatRuntimeException;
+import io.confluent.copycat.hdfs.SequenceFile.Reader;
+import io.confluent.copycat.hdfs.SequenceFile.Writer;
+
+public class WAL {
+
+  private static final Logger log = LoggerFactory.getLogger(WAL.class);
+  private SequenceFile.Writer writer = null;
+  private SequenceFile.Reader reader = null;
+  private FileSystem fs = null;
+  private String logFile = null;
+  private Configuration conf = null;
+  private String topicsDir = null;
+  private String url = null;
+  private TopicPartition topicPart = null;
+  private long sleepIntervalMills = 5000L;
+  private final String leaseException =
+      "org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException";
+
+  public WAL(Configuration conf, String url, String topicsDir, TopicPartition topicPart) {
+    try {
+      this.conf = conf;
+      this.topicsDir = topicsDir;
+      this.topicPart = topicPart;
+      this.url = url;
+      fs = FileSystem.get(URI.create(url), conf);
+      logFile = FileUtils.logFileName(url, topicsDir, topicPart);
+    } catch (IOException e) {
+      throw new CopycatRuntimeException(e);
+    }
+  }
+
+  public void append(String tempFile, String committedFile) throws InterruptedException {
+    while (true) {
+      try {
+        if (writer == null) {
+          writer = SequenceFile.createWriter(conf,
+                                             Writer.file(new Path(logFile)),
+                                             Writer.keyClass(Text.class),
+                                             Writer.valueClass(Text.class));
+        }
+        writer.append(new Text(tempFile), new Text(committedFile));
+        writer.hsync();
+        break;
+      } catch (RemoteException e) {
+        if (e.getClassName().equals(leaseException)) {
+          sleepIntervalMills = sleepIntervalMills * 2;
```

Comment:
The sleep interval should probably eventually be capped at some maximum amount of time (at some point you're going to end up killing your group membership anyway if this thread blocks for so long). sleepIntervalMills is never reset either, so even if you acquire the lease, if you later lose it and need to reacquire your timeout will still be at the old, increased value.

Finally, I think eventually we should give up if we really can't acquire it. Right now we'll loop forever if we keep getting lease exceptions.